test/test_custom_ops.py

```diff
@@ -2460,6 +2467,27 @@ def foo3(x: Tensor, *, y: int, z: List[Tensor]) -> Tensor:
                     lambda info, in_dims, x, *, y: (x, 0),
                 )
 
+    def test_dict_input(self):
+        @torch.library.custom_op("mylib::foo", mutates_args=())
+        def foo(d: dict, t: torch.Tensor) -> torch.Tensor:
+            return torch.sin(d["x"] - d["y"] + t)
+
+        d = {"x": torch.randn(2, 3), "y": torch.randn(2, 3)}
+        t = torch.randn(2, 3)
+        y = torch.ops.mylib.foo(d, t)
+        self.assertEqual(y, torch.sin(d["x"] - d["y"] + t))
+
+    def test_dataclass_input(self):
+        torch.utils._pytree.register_dataclass(Point)
+
+        @torch.library.custom_op("mylib::foo", mutates_args=())
+        def foo(a: Point) -> torch.Tensor:
+            return torch.sqrt(torch.sum((a.x - a.y) ** 2))
+
+        x = Point(x=torch.randn(2, 3), y=torch.randn(2, 3))
+        y = torch.ops.mylib.foo(x)
+        self.assertEqual(y, torch.sqrt(torch.sum((x.x - x.y) ** 2)))
```

Comment:
Some high-level comments:
1) Let's wait until after the branch cut (Monday) to merge this, assuming its ready before then. We don't want this feature to be partially in PyTorch 2.7.
2) eager-mode performance is pretty important. Can you do some benchmarking comparing e.g. custom ops with dict input (uses the new path) to custom ops with list inputs (uses the C++ dispatcher)?

Reply:
Agree! I'll do the perf benchmark while you conduct code review.